lib/activerecord_spanner_adapter/connection.rb

```diff
@@ -197,6 +197,13 @@ def run_batch
 
     def execute_query sql, params: nil, types: nil
       if params
+        # First process and remove any hints in the params that indicate that
+        # a different read staleness should be used than the default.
+        staleness_hint = params.find { |p| p[1].is_a? Arel::Visitors::StalenessHint }
```

Comment:
We need to let this pass through the binds / parameters because some of these functions are called by ActiveRecord. The process is generally as follows:
1. A read/query operation like for example `Singer.find_by id: 1` is executed by the application.
2. ActiveRecord translates this operation to a SQL query and an array of binds. These two are the only values that are expected by ActiveRecord, and also the only two values that will be passed on to the provider.
3. The provider must execute the SQL and binds (parameters) that is passed in by ActiveRecord.

But you're right that this might not be the best place to do the conversion, as this method is not directly called by ActiveRecord. Instead, that is the `execute` method in `DatabaseStatements`. I've moved the translation code to that class.